### DISCUSSION TITLE
How to verify the auth token?

### DISCUSSION THREAD

**Discussion by FenLiaoYu** [pocketbase/pocketbase#1611] [Q&A]
I want to send pocketbase's JWT token to my API server, how to verify it?
Any way to use pocketbase js SDK like firebase admin SDK?

**Comment by FenLiaoYu** [comment]
Found #725 after this question created 🤡
Maybe works, trying.

**Comment by ganigeorgiev** [comment]
> I want to send pocketbase's JWT token to my API server, how to verify it?

There is no dedicated "verify"-only endpoint but you can call `authRefresh()` (`POST 
/api/collections/users/auth-refresh`).
It will throw an error if the current `pb.authStore.token` is expired or invalid.

> Any way to use pocketbase js SDK like firebase admin SDK?

You can [authenticate as admin](https://pocketbase.io/docs/authentication/#authenticate-as-admin). All Admins can modify and access everything and API rules don't apply for them.

```js
await pb.admins.authWithPassword("test@example.com", "1234567890");

const result = await pb.collection("example").getList(1, 30);
```

**Reply by FenLiaoYu** [reply]
Thank you! I have another doubt.
Will `authRefresh()` change the auth state of the client? Its name makes me feel it will change the state of the client. Because I request from the server, not the client,

**Reply by ganigeorgiev** [reply]
`authRefresh()` will send the existing token and will fetch a new one. It will update the local `pb.authStore` state.
If you don't want to update the local state, you can create a new JS SDK instance OR send manually the POST request with `fetch` (you'll have to pass only the `Authorization: TOKEN` header).

**Comment by jakoblo** [comment]
After browsing the issues and discussions for a while i see that there are several people who want to use pocketbase authentication tokens for an external api or an already existing backend.

Even though I see that this is certainly not the original purpose of pocketbase there are to my knowledge a couple of ways to accomplish this. 

1. calling the refresh-token endpoint to "verify" a pocketbase token. (this of course means some additional traffic that pocketbase has to deal with)
2. using a hook to add a second token to the auth response what could be used to authenticate on a external system.

The second method could be solved using the provided onRecordAuthRequest Hook.
```javascript
	app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
		e.HttpContext.Response().Header().Set("X-MyCustom-Token", "the-token")
		e.HttpContext.SetCookie(&http.Cookie{})
		return nil
	})
```
I am not sure if this method is good and secure, but you could send your own jwt in the header or cookie which can be used to authenticate some external api. It would even be possible to control the token duration via pocketbase. A user password reset in pocketbase would not invalidate the custom token though. 

I like the way authentication is solved in pocketbase because it is very simple but perhaps my thoughts help someone and comments are very welcome

**Comment by Elliott-Green** [comment]
I'm facing the same issue now @ganigeorgiev, wondering if you have any advice.

I'm looking [here](https://pocketbase.io/docs/api-records) for auth refresh and the original question was the JWT token is coming from SK into an external API, but I can't see anywhere you'd actually pass the JWT into. So auth refresh is pretty much exclusively useful for the SK side, rather than on the SDK usage externally? 

I've got a Sveltekit, Express, and Pocketbase application. 

I'm at the point whereby I'm stitching SK+PB and Express together. I'm wanting a pattern where I can send the JWT onwards to be authenticated by express logic to toggle between 'free' data and 'logged in' data being returned to SK depending if the token is 1) sent to express, 2) is valid.

Would a verify only endpoint be able to be added? I might be able to test it?

**Reply by ganigeorgiev** [reply]
>  So auth refresh is pretty much exclusively useful for the SK side, rather than on the SDK usage externally?

I'm not sure that I understand the question. You can load an external token in the `pb.authStore` using `save`:

```js
pb.authStore.save(yourToken, null);

await pb.collection("users").authRefresh()
```

> Would a verify only endpoint be able to be added?

It is in my todo to consider it, but I'm still not sure how useful it would be since to some extend it will duplicate what `authRefresh` do and for now remains a very low priority.

**Reply by Elliott-Green** [reply]
Hey. Sorry I worded that badly.

Imagine you take the pocketbase JWT from `locals?.pb?.authStore?.token` and you pass onwards to another service. How does the second service verify that the pocketbase JWT is actually a valid session, additionally, can it tell me what users token that is?

**Reply by ganigeorgiev** [reply]
> @Elliott-Green  How does the second service verify that the pocketbase JWT is actually a valid session, additionally, can it tell me what users token that is?

The second service should call `POST  /api/collections/users/auth-refresh` with `Authorization: YOUR_TOKEN` header (or if you use the SDK use the code from the above comment). On valid and non-expired token, this will return an up-to-date user data and a token with new `exp` claim.

**Reply by Elliott-Green** [reply]
Thanks @ganigeorgiev, I got it working. Left a solution for anyone else reading this in the future.

```ts
import axios from 'axios';
const pocketbase_uri = process.env.POCKETBASE_URI;

const response = await axios.post(pocketbase_uri + '/api/collections/users/auth-refresh', {}, {
    headers: {
    'Content-Type': 'application/json',
    'Authorization': key
    }
});
console.log(JSON.stringify(response.data));
```

**Reply by ganigeorgiev** [reply]
@Elliott-Green Please note that you don't need an external HTTP client (axios in your case) just to make the request. If you are already loading the [JS SDK](https://github.com/pocketbase/js-sdk), you can use the code sample from https://github.com/pocketbase/pocketbase/discussions/1611#discussioncomment-5186938:

```js
pb.authStore.save(yourToken, null);

const authData = await pb.collection("users").authRefresh()
```

**Comment by ArslanT57** [comment]
Hi,
I am working with PB and Sveltekit. I have set up hooks to restrict access to visitors who are not authenticated. Following is the hooks.server.ts file:

```
import PocketBase from "pocketbase"
import { serializedNonPOJO } from "$lib/utils";
import { redirect, type Handle, error } from "@sveltejs/kit";

export const handle = async ({event, resolve}) => {
    event.locals.pb = new PocketBase("http://127.0.0.1:8090")

    // Auth store update
    event.locals.pb.authStore.loadFromCookie(event.request.headers.get("cookie") || "")

    const userIsAuthenticated = event.locals.pb.authStore.isValid
    const userGoingToLogin = event.url.pathname.startsWith("/login")
    const userForgotPassword = event.url.pathname.startsWith("/forgot-password")

    if (userIsAuthenticated) {
        event.locals.user = serializedNonPOJO(event.locals.pb.authStore.model)
        console.log(event.locals.user)
    } else if (!(userGoingToLogin || userForgotPassword)){
        event.locals.user = undefined
        throw redirect(303, "/login")
    } else {
        event.locals.user = undefined
    }

    const response = await resolve(event)

    response.headers.set("set-cookie", event.locals.pb.authStore.exportToCookie({secure: true}))
    return response
}
```
This code basically checks the request for auth cookie and updates the event.local.pb auth store. The 'local' object is then passed to load and actions on the page accessed, basic Sveltekit. Now, the cookie looks like this: (Authstore, locals.user in the code is the same just without token) 

```
{
	"token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2ODc3OTg4NjEsImlkIjoienQ0ZmlpbnA4eTNmOGtoIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.uTjyEiXuryRwPev8vqOZl8jMvQ_mN7hCYxG57ztlDdo",
	"model": {
		"avatar": "",
		"collectionId": "_pb_users_auth_",
		"collectionName": "users",
		"created": "2023-06-12 13:43:07.770Z",
		"departrments": ["admissions", "reports", "accounts", "classes"],
		"email": "REDACTED",
		"emailVisibility": false,
		"id": "zt4fiinp8y3f8kh",
		"name": "",
		"updated": "2023-06-12 16:58:57.099Z",
		"username": "users85485",
		"verified": true,
		"expand": {}
	}
}

```

What's troubling me is that user can easily just modify this cookie and the server will eat it up no complaints. In the department field above there are supposed to only three values but here they are four because I changed the cookie and now I have access to that department as well. That is because I was using the department field to trim the dashboard of my app to only show the links and pages related to the department a person is in. 

**Questions**
Is my current approach secure? I don't think it is because user can just willy nilly change their department. What I can think of to solve it is have a separate table to store the permissions and departments. But that will mean having to access the databse, PB, on every request. Same as authRefresh I think. Won't that overwhelm the PB server unnecessarliy and waste resources?

The other way I can think of is having a writeable store and store the auth token in that. Then store the key to that in the cookie. Other things remain the same.

Is there a better way to do authentication and later on authorization (same as I was trying to do with department field) more securely and without exposing this much data to the user.

Thanks

**Reply by ganigeorgiev** [reply]
If you want to safely trust the loaded cookie data you have to call `authRefresh()`. On valid token it will automatically update the AuthStore with the latest user data and a token with renewed `exp` claim, otherwise throw an error. For a SvelteKit example you can also check https://github.com/pocketbase/js-sdk#ssr-integration.

**Reply by ArslanT57** [reply]
Thanks, with authRefresh I can trust the cookie. BTW from your experience is this authentication method an efficient approach or should I look at something else?
Also, is there a way to strip the user information from the cookie and just use the auth token? Or will removing this info mean I have to fetch it on every request?

**Reply by ganigeorgiev** [reply]
>  from your experience is this authentication method an efficient approach or should I look at something else?

There is no other approach at the moment. In terms of efficiency, I haven't done extensive load tests yet but I don't think that the `authRefresh` call will impact the performance in any significant way (especially if the node app and PocketBase are on the same server or within the same network).

>  is there a way to strip the user information from the cookie and just use the auth token?

You can manually generate a cookie string with the `pb.authStore.token`. `exportToCookie` is just a small helper and it is not required to be used.

> Or will removing this info mean I have to fetch it on every request?

I'm not sure that I understand what do you mean. Even if you store only the token in a cookie, you still will have to send an `authRefresh()` call in order to be sure that you can trust it in your node application. You essentially have 2 backends - the SvelteKit server-side node app and PocketBase. The node application doesn't know anything about the token, the user data, etc. and in order to verify the loaded state, it needs to send request to PocketBase.

**Reply by flashedk** [reply]
Adding to this:
How do I verify on a route downstream that the user is still logged in?
Every solution i found so far basically stops after the login cookie is set.

```
{#if !loggedIn}
   <small>Hi you are not logged in</small>
   <button>login</button>
{:else}
   <small>Hi you are logged in</small>
   <button class="bg-red-900">logout</button>
{/if}
```

But how do I get the form to be loggedIn, because every time I console.log it, it shows false on the downstream route but true on the login page?
